File: google-drive/typescript/src/services/permissions.ts
/**
 * Permissions service for Google Drive API.
 *
 * Manages file and folder permissions, including sharing and access control.
 */

import {
  Permission,
  PermissionList,
  CreatePermissionRequest,
  UpdatePermissionRequest,
  ListPermissionsParams,
  GetPermissionParams,
  DeletePermissionParams,
} from '../types';

/**
 * HTTP transport interface for permission operations.
 */
export interface PermissionTransport {
  request<T>(url: string, options: RequestOptions): Promise<T>;
}

/**
 * Request options.
 */
export interface RequestOptions {
  method?: string;
  headers?: Record<string, string>;
  body?: any;
  params?: Record<string, any>;
}

/**
 * Permissions service interface.
 */
export interface PermissionsService {
  /**
   * Create a new permission (share a file).
   *
   * @param fileId - File ID
   * @param request - Permission creation request
   * @returns The created permission
   */
  create(fileId: string, request: CreatePermissionRequest): Promise<Permission>;

  /**
   * List permissions for a file.
   *
   * @param fileId - File ID
   * @param params - Optional list parameters
   * @returns Permission list with pagination
   */
  list(fileId: string, params?: ListPermissionsParams): Promise<PermissionList>;

  /**
   * Get a specific permission.
   *
   * @param fileId - File ID
   * @param permissionId - Permission ID
   * @param params - Optional parameters
   * @returns The permission
   */
  get(fileId: string, permissionId: string, params?: GetPermissionParams): Promise<Permission>;

  /**
   * Update a permission.
   *
   * @param fileId - File ID
   * @param permissionId - Permission ID
   * @param request - Update request
   * @returns The updated permission
   */
  update(
    fileId: string,
    permissionId: string,
    request: UpdatePermissionRequest
  ): Promise<Permission>;

  /**
   * Delete a permission (revoke access).
   *
   * @param fileId - File ID
   * @param permissionId - Permission ID
   * @param params - Optional delete parameters
   */
  delete(fileId: string, permissionId: string, params?: DeletePermissionParams): Promise<void>;
}

/**
 * Implementation of PermissionsService.
 */
export class PermissionsServiceImpl implements PermissionsService {
  private readonly baseUrl = 'https://www.googleapis.com/drive/v3';

  constructor(private transport: PermissionTransport) {}

  async create(fileId: string, request: CreatePermissionRequest): Promise<Permission> {
    const params: any = {};
    if (request.sendNotificationEmail !== undefined) {
      params.sendNotificationEmail = request.sendNotificationEmail;
    }
    if (request.emailMessage) {
      params.emailMessage = request.emailMessage;
    }
    if (request.transferOwnership) {
      params.transferOwnership = request.transferOwnership;
    }

    return this.transport.request<Permission>(
      `${this.baseUrl}/files/${fileId}/permissions`,
      {
        method: 'POST',
        body: {
          role: request.role,
          type: request.type,
          emailAddress: request.emailAddress,
          domain: request.domain,
          allowFileDiscovery: request.allowFileDiscovery,
          expirationTime: request.expirationTime,
          view: request.view,
        },
        params,
      }
    );
  }

  async list(fileId: string, params?: ListPermissionsParams): Promise<PermissionList> {
    return this.transport.request<PermissionList>(
      `${this.baseUrl}/files/${fileId}/permissions`,
      {
        method: 'GET',
        params: params as any,
      }
    );
  }

  async get(fileId: string, permissionId: string, params?: GetPermissionParams): Promise<Permission> {
    return this.transport.request<Permission>(
      `${this.baseUrl}/files/${fileId}/permissions/${permissionId}`,
      {
        method: 'GET',
        params: params as any,
      }
    );
  }

  async update(
    fileId: string,
    permissionId: string,
    request: UpdatePermissionRequest
  ): Promise<Permission> {
    return this.transport.request<Permission>(
      `${this.baseUrl}/files/${fileId}/permissions/${permissionId}`,
      {
        method: 'PATCH',
        body: request,
      }
    );
  }

  async delete(fileId: string, permissionId: string, params?: DeletePermissionParams): Promise<void> {
    await this.transport.request<void>(
      `${this.baseUrl}/files/${fileId}/permissions/${permissionId}`,
      {
        method: 'DELETE',
        params: params as any,
      }
    );
  }
}

/**
 * Mock implementation for testing.
 */
export class MockPermissionsService implements PermissionsService {
  private permissions = new Map<string, Map<string, Permission>>();
  private nextId = 1;

  async create(fileId: string, request: CreatePermissionRequest): Promise<Permission> {
    if (!this.permissions.has(fileId)) {
      this.permissions.set(fileId, new Map());
    }

    const permission: Permission = {
      kind: 'drive#permission',
      id: `permission-${this.nextId++}`,
      type: request.type,
      role: request.role,
      emailAddress: request.emailAddress,
      domain: request.domain,
      allowFileDiscovery: request.allowFileDiscovery,
      expirationTime: request.expirationTime,
      view: request.view,
    };

    this.permissions.get(fileId)!.set(permission.id, permission);
    return permission;
  }

  async list(fileId: string, params?: ListPermissionsParams): Promise<PermissionList> {
    const filePermissions = this.permissions.get(fileId);
    const permissions = filePermissions ? Array.from(filePermissions.values()) : [];

    return {
      kind: 'drive#permissionList',
      permissions,
    };
  }

  async get(fileId: string, permissionId: string, params?: GetPermissionParams): Promise<Permission> {
    const filePermissions = this.permissions.get(fileId);
    if (!filePermissions) {
      throw new Error(`File not found: ${fileId}`);
    }

    const permission = filePermissions.get(permissionId);
    if (!permission) {
      throw new Error(`Permission not found: ${permissionId}`);
    }

    return permission;
  }

  async update(
    fileId: string,
    permissionId: string,
    request: UpdatePermissionRequest
  ): Promise<Permission> {
    const permission = await this.get(fileId, permissionId);
    Object.assign(permission, request);
    return permission;
  }

  async delete(fileId: string, permissionId: string, params?: DeletePermissionParams): Promise<void> {
    const filePermissions = this.permissions.get(fileId);
    if (filePermissions) {
      filePermissions.delete(permissionId);
    }
  }
}

/**
 * Create a mock permissions service.
 */
export function createMockPermissionsService(): PermissionsService {
  return new MockPermissionsService();
}

File: google-drive/rust/src/services/permissions.rs
//! Permission operations service.

use crate::client::GoogleDriveClient;
use crate::error::GoogleDriveResult;
use crate::types::*;
use serde::{Deserialize, Serialize};

/// Service for permission operations.
pub struct PermissionsService<'a> {
    client: &'a GoogleDriveClient,
}

impl<'a> PermissionsService<'a> {
    /// Creates a new permissions service.
    pub fn new(client: &'a GoogleDriveClient) -> Self {
        Self { client }
    }

    /// Creates a new permission.
    pub async fn create(
        &self,
        file_id: &str,
        request: CreatePermissionRequest,
    ) -> GoogleDriveResult<Permission> {
        let path = format!("/files/{}/permissions", file_id);
        self.client.post(&path, &request).await
    }

    /// Lists permissions for a file.
    pub async fn list(
        &self,
        file_id: &str,
        params: Option<ListPermissionsParams>,
    ) -> GoogleDriveResult<PermissionList> {
        let path = format!("/files/{}/permissions", file_id);
        if let Some(p) = params {
            self.client.get_with_params(&path, &p).await
        } else {
            self.client.get(&path).await
        }
    }

    /// Gets a specific permission.
    pub async fn get(
        &self,
        file_id: &str,
        permission_id: &str,
        params: Option<GetPermissionParams>,
    ) -> GoogleDriveResult<Permission> {
        let path = format!("/files/{}/permissions/{}", file_id, permission_id);
        if let Some(p) = params {
            self.client.get_with_params(&path, &p).await
        } else {
            self.client.get(&path).await
        }
    }

    /// Updates a permission.
    pub async fn update(
        &self,
        file_id: &str,
        permission_id: &str,
        request: UpdatePermissionRequest,
    ) -> GoogleDriveResult<Permission> {
        let path = format!("/files/{}/permissions/{}", file_id, permission_id);
        self.client.patch(&path, &request).await
    }

    /// Deletes a permission.
    pub async fn delete(
        &self,
        file_id: &str,
        permission_id: &str,
        params: Option<DeletePermissionParams>,
    ) -> GoogleDriveResult<()> {
        let mut path = format!("/files/{}/permissions/{}", file_id, permission_id);
        if let Some(p) = params {
            if p.supports_all_drives {
                path.push_str("?supportsAllDrives=true");
            }
        }
        self.client.delete(&path).await
    }
}

// Request/Response types

/// Request to create a permission.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreatePermissionRequest {
    pub role: String,
    #[serde(rename = "type")]
    pub permission_type: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub email_address: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub domain: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub allow_file_discovery: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub expiration_time: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub send_notification_email: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub email_message: Option<String>,
}

/// Parameters for listing permissions.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct ListPermissionsParams {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub page_size: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub page_token: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub supports_all_drives: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub fields: Option<String>,
}

/// Parameters for getting a permission.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct GetPermissionParams {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub supports_all_drives: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub fields: Option<String>,
}

/// Request to update a permission.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct UpdatePermissionRequest {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub role: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub expiration_time: Option<String>,
}

/// Parameters for deleting a permission.
#[derive(Debug, Clone, Default)]
pub struct DeletePermissionParams {
    pub supports_all_drives: bool,
}
